File: src/gleam_mutants/gleam_mutants_platform_ffi.erl
%% SPDX-FileCopyrightText: 2026 gleam_mutants contributors
%% SPDX-License-Identifier: MIT OR Apache-2.0

-module(gleam_mutants_platform_ffi).
-include_lib("kernel/include/file.hrl").
-export([arguments/0, env/1, current_directory/0, temporary_directory/0,
         cache_directory/0, cpu_count/0, now_milliseconds/0, monotonic_milliseconds/0,
         resolve_path/1, architecture/0, environment/0, random_nonce/0,
         delete_tree/1,
         acquire_lock/4, release_lock/2, process_id/0,
         os_name/0, is_tty/0, is_reparse_point/1, exit/1, run_process/5, run_process_batch/2]).

delete_tree(Path) -> delete_tree(unicode:characters_to_list(Path), 40).

delete_tree(Path, Attempts) ->
    case delete_tree_once(Path) of
        ok -> <<>>;
        {error, enoent} -> <<>>;
        {error, _} when Attempts > 0 ->
            timer:sleep(50),
            delete_tree(Path, Attempts - 1);
        {error, Reason} ->
            unicode:characters_to_binary(io_lib:format("~tp", [Reason]))
    end.

delete_tree_once(Path) ->
    case file:read_link_info(Path) of
        {error, enoent} -> ok;
        {error, Reason} -> {error, Reason};
        {ok, Info} when Info#file_info.type =:= symlink ->
            delete_symlink(Path);
        {ok, Info} when Info#file_info.type =:= directory ->
            case file:list_dir(Path) of
                {ok, Names} ->
                    case delete_children(Path, Names) of
                        ok -> file:del_dir(Path);
                        Error -> Error
                    end;
                {error, Reason} -> {error, Reason}
            end;
        {ok, _} -> file:delete(Path)
    end.

delete_symlink(Path) ->
    case file:delete(Path) of
        {error, eperm} -> file:del_dir(Path);
        {error, eisdir} -> file:del_dir(Path);
        Result -> Result
    end.

delete_children(_Path, []) -> ok;
delete_children(Path, [Name | Rest]) ->
    case delete_tree_once(filename:join(Path, Name)) of
        ok -> delete_children(Path, Rest);
        Error -> Error
    end.

arguments() ->
    Raw = init:get_plain_arguments(),
    ScriptResult = try escript:script_name() of
        Value -> {ok, Value}
    catch
        _:_ -> error
    end,
    Args = case ScriptResult of
        {ok, Script} when is_list(Script) ->
            case Raw of
[First | Rest] when First =:= Script ->
                    case filelib:is_regular(First) of
                        true -> Rest;
                        false -> Raw
                    end;
                _ -> Raw
            end;
        _ -> Raw
    end,
    [unicode:characters_to_binary(A) || A <- Args].

env(Name) ->
    case os:getenv(unicode:characters_to_list(Name)) of
        false -> <<>>;
        Value -> unicode:characters_to_binary(Value)
    end.

current_directory() ->
    {ok, Directory} = file:get_cwd(),
    unicode:characters_to_binary(Directory).

temporary_directory() ->
    case os:type() of
        {win32, _} -> first_env([<<"TEMP">>, <<"TMP">>], <<".">>);
        _ -> first_env([<<"TMPDIR">>], <<"/tmp">>)
    end.

cache_directory() ->
    case os:type() of
        {win32, _} -> first_env([<<"LOCALAPPDATA">>], temporary_directory());
        {unix, darwin} -> <<(first_env([<<"HOME">>], <<".">>))/binary, "/Library/Caches">>;
        _ -> first_env([<<"XDG_CACHE_HOME">>], <<(first_env([<<"HOME">>], <<".">>))/binary, "/.cache">>)
    end.

first_env([], Default) -> Default;
first_env([Name | Rest], Default) ->
    case env(Name) of <<>> -> first_env(Rest, Default); Value -> Value end.

cpu_count() ->
    case erlang:system_info(logical_processors_available) of
        unknown -> 1;
        Count when is_integer(Count), Count > 0 -> Count;
        _ -> 1
    end.

now_milliseconds() -> erlang:system_time(millisecond).
monotonic_milliseconds() -> erlang:monotonic_time(millisecond).
resolve_path(Path) -> unicode:characters_to_binary(filename:absname(unicode:characters_to_list(Path))).
architecture() -> unicode:characters_to_binary(erlang:system_info(system_architecture)).
environment() ->
    Entries = lists:sort(os:getenv()),
    unicode:characters_to_binary(lists:flatten([
        integer_to_list(length(Entry)) ++ ":" ++ Entry || Entry <- Entries
    ])).
random_nonce() ->
    Bytes = crypto:strong_rand_bytes(16),
    unicode:characters_to_binary(binary:encode_hex(Bytes, lowercase)).

acquire_lock(Path0, RunId, Started, WaitMs) ->
    Path = unicode:characters_to_list(Path0),
    ok = filelib:ensure_dir(Path),
    Token = random_nonce(),
    Deadline = erlang:monotonic_time(millisecond) + WaitMs,
    acquire_lock_loop(Path, Token, RunId, Started, Deadline).

acquire_lock_loop(Path, Token, RunId, Started, Deadline) ->
    case file:open(Path, [write, exclusive, binary]) of
        {ok, File} ->
            Metadata = <<Token/binary, "\n", (integer_to_binary(process_id()))/binary,
                         "\n", RunId/binary, "\n", (integer_to_binary(Started))/binary, "\n">>,
            Result = case file:write(File, Metadata) of
                ok -> file:sync(File);
                Error -> Error
            end,
            ok = file:close(File),
            case Result of
                ok -> <<"ok:", Token/binary>>;
                {error, Reason} ->
                    _ = file:delete(Path),
                    unicode:characters_to_binary(io_lib:format("error:could not write workspace lock: ~p", [Reason]))
            end;
        {error, eexist} ->
            case read_lock(Path) of
                {ok, _OldToken, Pid, _OldRunId, _OldStarted} ->
                    case lock_process_alive(Pid) of
                        false -> _ = file:delete(Path), acquire_lock_loop(Path, Token, RunId, Started, Deadline);
                        true -> wait_for_lock(Path, Token, RunId, Started, Deadline)
                    end;
                error -> wait_for_lock(Path, Token, RunId, Started, Deadline)
            end;
        {error, Reason} ->
            unicode:characters_to_binary(io_lib:format("error:could not create workspace lock: ~p", [Reason]))
    end.

wait_for_lock(Path, Token, RunId, Started, Deadline) ->
    case erlang:monotonic_time(millisecond) >= Deadline of
        true ->
            case read_lock(Path) of
                {ok, _, Pid, OldRunId, OldStarted} ->
                    <<"error:workspace is locked by pid ", (integer_to_binary(Pid))/binary,
                      ", run ", OldRunId/binary, ", started ", OldStarted/binary>>;
                error -> <<"error:workspace lock is busy">>
            end;
        false -> timer:sleep(50), acquire_lock_loop(Path, Token, RunId, Started, Deadline)
    end.

read_lock(Path) ->
    case file:read_file(Path) of
        {ok, Data} ->
            case binary:split(Data, <<"\n">>, [global]) of
                [Token, Pid, RunId, Started | _] ->
                    try {ok, Token, binary_to_integer(Pid), RunId, Started}
                    catch _:_ -> error end;
                _ -> error
            end;
        _ -> error
    end.

lock_process_alive(Pid) ->
    Command = case os:type() of
        {win32, _} -> "tasklist /FI \"PID eq " ++ integer_to_list(Pid) ++ "\" /NH";
        _ -> "kill -0 " ++ integer_to_list(Pid) ++ " 2>/dev/null && echo alive"
    end,
    string:find(os:cmd(Command), case os:type() of {win32, _} -> integer_to_list(Pid); _ -> "alive" end) =/= nomatch.

release_lock(Path0, Token) ->
    Path = unicode:characters_to_list(Path0),
    case read_lock(Path) of
        {ok, Token, _, _, _} ->
            case file:delete(Path) of
                ok -> <<>>;
                {error, Reason} -> unicode:characters_to_binary(io_lib:format("could not release workspace lock: ~p", [Reason]))
            end;
        {ok, _, _, _, _} -> <<"workspace lock ownership changed">>;
        error -> <<"could not read workspace lock during release">>
    end.
process_id() -> list_to_integer(os:getpid()).

os_name() ->
    case os:type() of
        {win32, _} -> <<"windows">>;
        {unix, darwin} -> <<"darwin">>;
        {unix, _} -> <<"linux">>;
        _ -> <<"unknown">>
    end.

is_tty() ->
    case io:columns() of {ok, _} -> true; _ -> false end.

is_reparse_point(Path) ->
    case file:read_link_info(unicode:characters_to_list(Path)) of
        {ok, Info} when Info#file_info.type =:= symlink -> true;
        _ -> false
    end.
exit(Code) -> erlang:halt(Code).

run_process(Executable0, Args0, WorkingDirectory, Environment, Timeout) ->
    Executable = resolve_executable(Executable0),
    Args = [unicode:characters_to_list(A) || A <- Args0],
    Env = [{unicode:characters_to_list(K), unicode:characters_to_list(V)} || {K, V} <- Environment],
    Options = [binary, exit_status, eof, use_stdio, stderr_to_stdout, hide,
               {args, Args}, {cd, unicode:characters_to_list(WorkingDirectory)}, {env, Env}],
    Result = try erlang:open_port({spawn_executable, Executable}, Options) of
        Port -> collect(Port, Timeout, erlang:monotonic_time(millisecond), <<>>, undefined, false)
    catch
        Class:Reason -> {-2, <<>>, unicode:characters_to_binary(io_lib:format("~p:~p", [Class, Reason])), false}
    end,
    preserve_interrupt(Result).

preserve_interrupt({130, _, _, _}) -> erlang:halt(130);
preserve_interrupt(Result) -> Result.

run_process_batch([], _Jobs) -> [];
run_process_batch(Requests, Jobs) ->
    Count = erlang:min(erlang:max(1, Jobs), length(Requests)),
    {Batch, Rest} = lists:split(Count, Requests),
    Parent = self(),
    Refs = [begin
        Ref = make_ref(),
        spawn(fun() ->
            Started = erlang:monotonic_time(millisecond),
            {Executable, Args, WorkingDirectory, Environment, Timeout} = Request,
            {Status, Stdout, Stderr, TimedOut} =
                run_process(Executable, Args, WorkingDirectory, Environment, Timeout),
            Duration = erlang:monotonic_time(millisecond) - Started,
            Parent ! {Ref, {Status, Stdout, Stderr, TimedOut, Duration}}
        end),
        Ref
    end || Request <- Batch],
    Results = [receive {Ref, Result} -> Result end || Ref <- Refs],
    Results ++ run_process_batch(Rest, Jobs).
resolve_executable(Name) ->
    NameList = unicode:characters_to_list(Name),
    case filename:pathtype(NameList) of
        absolute -> NameList;
        _ -> case os:find_executable(NameList) of false -> NameList; Path -> Path end
    end.

collect(Port, Timeout, Started, Output, Status, Eof) ->
    Elapsed = erlang:monotonic_time(millisecond) - Started,
    Remaining0 = erlang:max(0, Timeout - Elapsed),
    Remaining = case Status of undefined -> Remaining0; _ -> erlang:min(Remaining0, 5000) end,
    receive
        {Port, {data, Chunk}} ->
            collect(Port, Timeout, Started, bounded_append(Output, Chunk), Status, Eof);
        {Port, {exit_status, Code}} when Eof =:= true ->
            {Code, Output, <<>>, false};
        {Port, {exit_status, Code}} ->
            collect(Port, Timeout, Started, Output, Code, Eof);
        {Port, eof} when Status =/= undefined ->
            {Status, Output, <<>>, false};
        {Port, eof} ->
            collect(Port, Timeout, Started, Output, Status, true);
        {'EXIT', Port, _} -> {status_or_default(Status), Output, <<>>, false}
    after Remaining ->
        case Status of
            undefined ->
                kill_port_tree(Port),
                try erlang:port_close(Port) catch _:_ -> ok end,
                {-1, Output, <<"process timed out">>, true};
            _ -> {Status, Output, <<>>, false}
        end
    end.

bounded_append(Output, Chunk) ->
    Combined = <<Output/binary, Chunk/binary>>,
    Limit = 131072,
    Half = Limit div 2,
    case byte_size(Combined) =< Limit of
        true -> Combined;
        false ->
            Head = binary:part(Combined, 0, Half),
            Tail = binary:part(Combined, byte_size(Combined) - Half, Half),
            <<Head/binary, "\n... output truncated ...\n", Tail/binary>>
    end.

status_or_default(undefined) -> -1;
status_or_default(Status) -> Status.

kill_port_tree(Port) ->
    case erlang:port_info(Port, os_pid) of
        {os_pid, Pid} -> kill_tree(Pid);
        _ -> ok
    end.

kill_tree(Pid) ->
    case os:type() of
        {win32, _} ->
            os:cmd("taskkill /PID " ++ integer_to_list(Pid) ++ " /T /F >NUL 2>&1"), ok;
        _ ->
            Descendants = descendants(Pid),
            signal_all(Descendants ++ [Pid], "-TERM"),
            timer:sleep(250),
            signal_all(Descendants ++ [Pid], "-KILL")
    end.

signal_all(Pids, Signal) ->
    lists:foreach(fun(Pid) -> os:cmd("kill " ++ Signal ++ " " ++ integer_to_list(Pid) ++ " 2>/dev/null") end, Pids), ok.

descendants(Root) ->
    Lines = string:split(os:cmd("ps -eo pid=,ppid="), "\n", all),
    Pairs = lists:filtermap(fun parse_pair/1, Lines),
    descendants_loop([Root], Pairs, []).

parse_pair(Line) ->
    case string:tokens(Line, " \t") of
        [Pid, Parent] ->
            try {true, {list_to_integer(Pid), list_to_integer(Parent)}} catch _:_ -> false end;
        _ -> false
    end.

descendants_loop([], _Pairs, Found) -> lists:reverse(Found);
descendants_loop([Parent | Queue], Pairs, Found) ->
    Children = [Pid || {Pid, Ppid} <- Pairs, Ppid =:= Parent, not lists:member(Pid, Found)],
    descendants_loop(Queue ++ Children, Pairs, Children ++ Found).

File: src/gleam_mutants/gleam_mutants_platform_ffi.mjs
// SPDX-FileCopyrightText: 2026 gleam_mutants contributors
// SPDX-License-Identifier: MIT OR Apache-2.0

import childProcess from "node:child_process";
import crypto from "node:crypto";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { toList } from "../gleam.mjs";

function listToArray(list) {
  return Array.from(list);
}

export function arguments$() {
  if (globalThis.Deno?.args) return toList([...globalThis.Deno.args]);
  return toList(process.argv.slice(2));
}

// Gleam escapes the reserved JS identifier in generated calls by using the
// exported source name, so also provide the unescaped property explicitly.
export { arguments$ as arguments };

export function env(name) {
  try {
    if (globalThis.Deno?.env) return globalThis.Deno.env.get(name) ?? "";
  } catch (_) {
    // Deno without env permission: fall through to the Node compatibility API.
  }
  return globalThis.process?.env?.[name] ?? "";
}

export function current_directory() {
  try { return globalThis.Deno?.cwd?.() ?? process.cwd(); }
  catch (_) { return process.cwd(); }
}

export function temporary_directory() {
  return os.tmpdir();
}

export function cache_directory() {
  if (process.platform === "win32") {
    return process.env.LOCALAPPDATA || path.join(os.homedir(), "AppData", "Local");
  }
  if (process.platform === "darwin") {
    return path.join(os.homedir(), "Library", "Caches");
  }
  return process.env.XDG_CACHE_HOME || path.join(os.homedir(), ".cache");
}

export function cpu_count() {
  return Math.max(1, os.availableParallelism?.() ?? os.cpus().length ?? 1);
}

export function now_milliseconds() { return Date.now(); }
export function monotonic_milliseconds() {
  return Math.trunc(globalThis.performance?.now?.() ?? Number(process.hrtime.bigint() / 1_000_000n));
}
export function resolve_path(value) { return path.resolve(value); }
export function architecture() { return globalThis.Deno?.build?.arch ?? process.arch; }
export function environment() {
  try {
    const entries = globalThis.Deno?.env ? [...globalThis.Deno.env.toObject ? Object.entries(globalThis.Deno.env.toObject()) : []] : Object.entries(process.env);
    return entries.map(([key, value]) => `${key.length}:${key}${String(value ?? "").length}:${value ?? ""}`).sort().join("");
  } catch (_) {
    return Object.entries(globalThis.process?.env ?? {}).map(([key, value]) => `${key.length}:${key}${String(value ?? "").length}:${value ?? ""}`).sort().join("");
  }
}
export function random_nonce() {
  if (globalThis.crypto?.getRandomValues) {
    const bytes = globalThis.crypto.getRandomValues(new Uint8Array(16));
    return [...bytes].map(byte => byte.toString(16).padStart(2, "0")).join("");
  }
  return crypto.randomBytes(16).toString("hex");
}
export function delete_tree(value) {
  try {
    fs.rmSync(path.normalize(value), {
      recursive: true,
      force: true,
      maxRetries: 40,
      retryDelay: 50,
    });
    return "";
  } catch (error) {
    return String(error?.message || error);
  }
}
function synchronousDelay(milliseconds) {
  Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, milliseconds);
}
function lockProcessAlive(pid) {
  try { process.kill(Number(pid), 0); return true; }
  catch (error) { return error?.code === "EPERM"; }
}
function parseLock(value) {
  const [token = "", pid = "", runId = "", started = ""] = String(value).split("\n");
  return { token, pid, runId, started };
}
export function acquire_lock(target, runId, startedMs, waitMs) {
  fs.mkdirSync(path.dirname(target), { recursive: true });
  const token = random_nonce();
  const deadline = Date.now() + waitMs;
  while (true) {
    try {
      const descriptor = fs.openSync(target, "wx", 0o600);
      try {
        fs.writeFileSync(descriptor, `${token}\n${process.pid}\n${runId}\n${startedMs}\n`, "utf8");
        fs.fsyncSync(descriptor);
      } finally { fs.closeSync(descriptor); }
      return `ok:${token}`;
    } catch (error) {
      if (error?.code !== "EEXIST") return `error:could not create workspace lock: ${error?.message || error}`;
      let existing = { pid: "unknown", runId: "unknown", started: "unknown" };
      try {
        existing = parseLock(fs.readFileSync(target, "utf8"));
        if (existing.pid && !lockProcessAlive(existing.pid)) {
          fs.unlinkSync(target);
          continue;
        }
      } catch (_) {}
      if (Date.now() >= deadline) {
        return `error:workspace is locked by pid ${existing.pid}, run ${existing.runId}, started ${existing.started}`;
      }
      synchronousDelay(50);
    }
  }
}
export function release_lock(target, token) {
  try {
    const existing = parseLock(fs.readFileSync(target, "utf8"));
    if (existing.token !== token) return "workspace lock ownership changed";
    fs.unlinkSync(target);
    return "";
  } catch (error) {
    return `could not release workspace lock: ${error?.message || error}`;
  }
}
export function process_id() { return process.pid; }
export function os_name() { return process.platform === "win32" ? "windows" : process.platform; }
export function is_tty() { return Boolean(process.stdout?.isTTY && process.stdin?.isTTY); }

export function is_reparse_point(target) {
  if (process.platform !== "win32") return false;
  try { return fs.lstatSync(target).isSymbolicLink(); }
  catch (_) { return false; }
}

export function exit(code) {
  if (globalThis.Deno) Deno.exit(code);
  process.exit(code);
}

function descendants(rootPid) {
  if (process.platform === "win32") return [];
  const result = childProcess.spawnSync("ps", ["-eo", "pid=,ppid="], { encoding: "utf8" });
  const rows = String(result.stdout || "").trim().split(/\r?\n/).map((line) => line.trim().split(/\s+/).map(Number));
  const found = [];
  const visit = (pid) => {
    for (const [child, parent] of rows) {
      if (parent === pid && !found.includes(child)) { found.push(child); visit(child); }
    }
  };
  visit(rootPid);
  return found.reverse();
}

function killTree(pid) {
  if (!pid) return;
  if (process.platform === "win32") {
    childProcess.spawnSync("taskkill", ["/PID", String(pid), "/T", "/F"], { windowsHide: true });
    return;
  }
  const pids = [...descendants(pid), pid];
  for (const signal of ["SIGTERM", "SIGKILL"]) {
    for (const item of pids) { try { process.kill(item, signal); } catch (_) {} }
    if (signal === "SIGTERM") Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, 250);
  }
}

const denoProcessWorkerSource = String.raw`
const delay = (milliseconds) => new Promise(resolve => setTimeout(resolve, milliseconds));
const OUTPUT_LIMIT = 64 * 1024;

async function readLimited(stream) {
  const reader = stream.getReader();
  const chunks = [];
  let size = 0;
  let omitted = 0;
  while (true) {
    const { value, done } = await reader.read();
    if (done) break;
    const remaining = Math.max(0, OUTPUT_LIMIT - size);
    if (remaining > 0) {
      const kept = value.subarray(0, remaining);
      chunks.push(kept);
      size += kept.length;
      omitted += value.length - kept.length;
    } else omitted += value.length;
  }
  const result = new Uint8Array(size);
  let offset = 0;
  for (const chunk of chunks) { result.set(chunk, offset); offset += chunk.length; }
  const text = new TextDecoder().decode(result);
  return omitted > 0 ? text + "\n... " + omitted + " bytes omitted ...\n" : text;
}

function descendants(rootPid, rows) {
  const found = [];
  const visit = (pid) => {
    for (const [child, parent] of rows) {
      if (parent === pid && !found.includes(child)) { found.push(child); visit(child); }
    }
  };
  visit(rootPid);
  return found.reverse();
}

async function killTree(pid) {
  if (Deno.build.os === "windows") {
    try {
      await new Deno.Command("taskkill", {
        args: ["/PID", String(pid), "/T", "/F"], stdout: "null", stderr: "null",
      }).output();
    } catch (_) {}
    return;
  }
  let pids = [pid];
  try {
    const output = await new Deno.Command("ps", {
      args: ["-eo", "pid=,ppid="], stdout: "piped", stderr: "null",
    }).output();
    const rows = new TextDecoder().decode(output.stdout).trim().split(/\r?\n/)
      .map(line => line.trim().split(/\s+/).map(Number));
    pids = [...descendants(pid, rows), pid];
  } catch (_) {}
  for (const signal of ["SIGTERM", "SIGKILL"]) {
    for (const item of pids) { try { Deno.kill(item, signal); } catch (_) {} }
    if (signal === "SIGTERM") await delay(250);
  }
}

self.onmessage = async ({ data }) => {
  const control = new Int32Array(data.shared, 0, 2);
  const destination = new Uint8Array(data.shared, 8);
  const finish = (value) => {
    const encoded = new TextEncoder().encode(JSON.stringify(value));
    const length = Math.min(encoded.length, destination.length);
    destination.set(encoded.subarray(0, length));
    Atomics.store(control, 1, length);
    Atomics.store(control, 0, 1);
    Atomics.notify(control, 0);
  };
  try {
    const child = new Deno.Command(data.executable, {
      args: data.arguments,
      cwd: data.workingDirectory || undefined,
      env: Object.fromEntries(data.environment),
      clearEnv: false,
      stdout: "piped",
      stderr: "piped",
    }).spawn();
    const stdoutPromise = readLimited(child.stdout);
    const stderrPromise = readLimited(child.stderr);
    const statusPromise = child.status;
    let timer;
    const timeoutPromise = new Promise(resolve => { timer = setTimeout(() => resolve(null), data.timeoutMs); });
    let status = await Promise.race([statusPromise, timeoutPromise]);
    let timedOut = false;
    if (status === null) {
      timedOut = true;
      await killTree(child.pid);
      status = await Promise.race([statusPromise, delay(1000).then(() => null)]);
    }
    clearTimeout(timer);
    const [stdout, stderr] = await Promise.all([stdoutPromise, stderrPromise]);
    finish([
      status?.code ?? (timedOut ? -1 : -2),
      stdout,
      stderr || (status ? "" : "process timed out"),
      timedOut,
    ]);
  } catch (error) {
    finish([-2, "", String(error?.message || error), false]);
  }
};
`;

function startDenoProcess(executable, arguments_, workingDirectory, environment, timeoutMs) {
  const shared = new SharedArrayBuffer(8 + 256 * 1024);
  const control = new Int32Array(shared, 0, 2);
  const workerUrl = URL.createObjectURL(new Blob([denoProcessWorkerSource], { type: "text/javascript" }));
  const worker = new Worker(workerUrl, { type: "module" });
  worker.postMessage({
    executable,
    arguments: arguments_,
    workingDirectory,
    environment,
    timeoutMs,
    shared,
  });
  return { shared, control, workerUrl, worker, timeoutMs };
}

function finishDenoProcess(handle) {
  try {
    const { shared, control, timeoutMs } = handle;
    const wait = Atomics.wait(control, 0, 0, timeoutMs + 5000);
    if (wait === "timed-out") return [-1, "", "process worker timed out", true];
    const length = Atomics.load(control, 1);
    const bytes = new Uint8Array(shared, 8, length);
    return JSON.parse(new TextDecoder().decode(bytes));
  } finally {
    handle.worker.terminate();
    URL.revokeObjectURL(handle.workerUrl);
  }
}

function runDenoProcess(executable, arguments_, workingDirectory, environment, timeoutMs) {
  return finishDenoProcess(startDenoProcess(executable, arguments_, workingDirectory, environment, timeoutMs));
}

const nodeBatchSource = String.raw`
const childProcess = require("node:child_process");
const requests = JSON.parse(process.env.GLEAM_MUTANTS_BATCH_REQUESTS);
const jobs = Math.max(1, Number(process.env.GLEAM_MUTANTS_BATCH_JOBS) || 1);
const activeChildren = new Set();
const OUTPUT_LIMIT = 64 * 1024;

class LimitedOutput {
  constructor() { this.chunks = []; this.size = 0; this.omitted = 0; }
  push(chunk) {
    const remaining = Math.max(0, OUTPUT_LIMIT - this.size);
    if (remaining > 0) {
      const kept = chunk.subarray(0, remaining);
      this.chunks.push(kept);
      this.size += kept.length;
      this.omitted += chunk.length - kept.length;
    } else this.omitted += chunk.length;
  }
  text() {
    const value = Buffer.concat(this.chunks, this.size).toString("utf8");
    return this.omitted > 0 ? value + "\n... " + this.omitted + " bytes omitted ...\n" : value;
  }
}

function descendants(rootPid) {
  if (process.platform === "win32") return [];
  const result = childProcess.spawnSync("ps", ["-eo", "pid=,ppid="], { encoding: "utf8" });
  const rows = String(result.stdout || "").trim().split(/\r?\n/)
    .map(line => line.trim().split(/\s+/).map(Number));
  const found = [];
  const visit = pid => {
    for (const [child, parent] of rows) {
      if (parent === pid && !found.includes(child)) { found.push(child); visit(child); }
    }
  };
  visit(rootPid);
  return found.reverse();
}

function killTree(pid) {
  if (!pid) return;
  if (process.platform === "win32") {
    childProcess.spawnSync("taskkill", ["/PID", String(pid), "/T", "/F"], { windowsHide: true });
    return;
  }
  const pids = [...descendants(pid), pid];
  for (const signal of ["SIGTERM", "SIGKILL"]) {
    for (const item of pids) { try { process.kill(item, signal); } catch (_) {} }
    if (signal === "SIGTERM") Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, 250);
  }
}

function interrupt(signal) {
  for (const pid of activeChildren) killTree(pid);
  process.exit(signal === "SIGINT" ? 130 : 143);
}

process.once("SIGINT", () => interrupt("SIGINT"));
process.once("SIGTERM", () => interrupt("SIGTERM"));

function run(request) {
  return new Promise(resolve => {
    const [executable, args, cwd, environment, timeoutMs] = request;
    const started = Date.now();
    const child = childProcess.spawn(executable, args, {
      cwd: cwd || undefined,
      env: { ...process.env, ...Object.fromEntries(environment) },
      windowsHide: true,
      stdio: ["ignore", "pipe", "pipe"],
    });
    if (child.pid) activeChildren.add(child.pid);
    const stdout = new LimitedOutput();
    const stderr = new LimitedOutput();
    let timedOut = false;
    let finished = false;
    child.stdout.on("data", chunk => stdout.push(chunk));
    child.stderr.on("data", chunk => stderr.push(chunk));
    const timer = setTimeout(() => {
      timedOut = true;
      killTree(child.pid);
    }, timeoutMs);
    const finish = (status, error = "") => {
      if (finished) return;
      finished = true;
      if (child.pid) activeChildren.delete(child.pid);
      clearTimeout(timer);
      resolve([
        status ?? (timedOut ? -1 : -2),
        stdout.text(),
        stderr.text() + error,
        timedOut,
        Date.now() - started,
      ]);
    };
    child.on("error", error => finish(-2, String(error.message || error)));
    child.on("close", code => finish(code));
  });
}

(async () => {
  const results = new Array(requests.length);
  let cursor = 0;
  async function worker() {
    while (true) {
      const index = cursor++;
      if (index >= requests.length) return;
      results[index] = await run(requests[index]);
    }
  }
  await Promise.all(Array.from({ length: Math.min(jobs, requests.length) }, worker));
  if (results.some(result => result[0] === 130)) process.exit(130);
  process.stdout.write(JSON.stringify(results));
})().catch(error => { process.stderr.write(String(error.stack || error)); process.exit(1); });
`;

export function run_process_batch(requestList, jobs) {
  const requests = listToArray(requestList).map(([executable, args, cwd, environment, timeoutMs]) => [
    executable,
    listToArray(args),
    cwd,
    listToArray(environment),
    timeoutMs,
  ]);
  if (requests.length === 0) return toList([]);
  if (globalThis.Deno) {
    const results = [];
    for (let cursor = 0; cursor < requests.length; cursor += Math.max(1, jobs)) {
      const wave = requests.slice(cursor, cursor + Math.max(1, jobs));
      const handles = wave.map(([executable, args, cwd, environment, timeoutMs]) =>
        startDenoProcess(executable, args, cwd, environment, timeoutMs));
      for (const handle of handles) {
        const started = performance.now();
        const result = finishDenoProcess(handle);
        results.push([...result, Math.max(0, Math.trunc(performance.now() - started))]);
      }
    }
    if (results.some(result => result[0] === 130)) Deno.exit(130);
    return toList(results);
  }
  const arguments_ = ["-e", nodeBatchSource];
  let status;
  let stdout;
  let stderr;
  const output = childProcess.spawnSync(process.execPath, arguments_, {
    encoding: "utf8",
    windowsHide: true,
    maxBuffer: 8 * 1024 * 1024,
    env: {
      ...process.env,
      GLEAM_MUTANTS_BATCH_REQUESTS: JSON.stringify(requests),
      GLEAM_MUTANTS_BATCH_JOBS: String(jobs),
    },
  });
  status = output.status ?? -2;
  stdout = String(output.stdout || "");
  stderr = String(output.stderr || output.error?.message || "");
  if (status !== 0) {
    if (status === 130) {
      if (globalThis.Deno) Deno.exit(130);
      process.exit(130);
    }
    return toList(requests.map(() => [-2, "", `batch runner failed: ${stderr}`, false, 0]));
  }
  return toList(JSON.parse(stdout));
}
export function run_process(executable, argumentList, workingDirectory, environment, timeoutMs) {
  const arguments_ = listToArray(argumentList);
  const overrides = listToArray(environment);
  if (globalThis.Deno) {
    const result = runDenoProcess(executable, arguments_, workingDirectory, overrides, timeoutMs);
    if (result[0] === 130) Deno.exit(130);
    return result;
  }
  const result = Array.from(run_process_batch(toList([
    [executable, toList(arguments_), workingDirectory, toList(overrides), timeoutMs],
  ]), 1))[0];
  return result.slice(0, 4);
}
